politeiad/backendv2/tstorebe/plugins/ticketvote/cmds.go

```diff
@@ -0,0 +1,2976 @@
+		commitmentAddr, ok := addrs[v.Ticket]
+		if !ok {
+			t := time.Now().Unix()
+			log.Errorf("cmdCastBallot: commitment addr not found %v: %v",
+				t, v.Ticket)
+			e := ticketvote.VoteErrorInternalError
+			receipts[k].Ticket = v.Ticket
+			receipts[k].ErrorCode = e
+			receipts[k].ErrorContext = fmt.Sprintf("%v: %v",
+				ticketvote.VoteErrors[e], t)
+			continue
+		}
+		if commitmentAddr.err != nil {
+			t := time.Now().Unix()
+			log.Errorf("cmdCastBallot: commitment addr error %v: %v %v",
+				t, v.Ticket, commitmentAddr.err)
+			e := ticketvote.VoteErrorInternalError
+			receipts[k].Ticket = v.Ticket
+			receipts[k].ErrorCode = e
+			receipts[k].ErrorContext = fmt.Sprintf("%v: %v",
+				ticketvote.VoteErrors[e], t)
+			continue
+		}
+		err = p.castVoteSignatureVerify(v, commitmentAddr.addr)
+		if err != nil {
+			e := ticketvote.VoteErrorSignatureInvalid
+			receipts[k].Ticket = v.Ticket
+			receipts[k].ErrorCode = e
+			receipts[k].ErrorContext = fmt.Sprintf("%v: %v",
+				ticketvote.VoteErrors[e], err)
+			continue
+		}
+	}
+
+	// The votes that have passed validation will be cast in batches of
+	// size batchSize. Each batch of votes is cast concurrently in order
+	// to accommodate the trillian log signer bottleneck. The log signer
+	// picks up queued leaves and appends them onto the trillian tree
+	// every xxx ms, where xxx is a configurable value on the log signer,
+	// but is typically a few hundred milliseconds. Lets use 200ms as an
+	// example. If we don't cast the votes in batches then every vote in
+	// the ballot will take 200 milliseconds since we wait for the leaf
+	// to be fully appended before considering the trillian call
+	// successful. A person casting hundreds of votes in a single ballot
+	// would cause UX issues for all the voting clients since the backend
+	// locks the record during any plugin write calls. Only one ballot
+	// can be cast at a time.
+	//
+	// The second variable that we must watch out for is the max trillian
+	// queued leaf batch size. This is also a configurable trillian value
+	// that represents the maximum number of leaves that can be waiting
+	// in the queue for all trees in the trillian instance. This value is
+	// typically around the order of magnitude of 1000s of queued leaves.
+	//
+	// The third variable that can cause errors is reaching the trillian
+	// datastore max connection limits. Each vote being cast creates a
+	// trillian connection. Overloading the trillian connections can
+	// cause max connection exceeded errors. The max allowed connections
+	// is a configurable trillian value, but should also be adjusted on
+	// the key-value store database itself as well.
+	//
+	// This is why a vote batch size of 10 was chosen. It is large enough
+	// to alleviate performance bottlenecks from the log signer interval,
+	// but small enough to still allow multiple records votes to be held
+	// concurrently without running into the queued leaf batch size limit.
+
```

Comment:
This is kinda of a terrible hack. I understand why it does what it does but is there no way to enhance this short of trillian itself? Can we, for example, use batched trillian writes instead?

Reply:
Not without allowing plugins to save batches of blob entries, which we really do not want to do.

Reply:
I'd like to explore this a bit more. Let's do a call. I feel like this will come and haunt us later.